nixos/iwd: add `networks` and `interfaces` option

Ma27 opened the pull request:
<!-- Nixpkgs has a lot of new incoming Pull Requests, but not enough people to review this constant stream. Even if you aren't a committer, we would appreciate reviews of other PRs, especially simple ones like package updates. Just testing the relevant package/service and leaving a comment saying what you tested, how you tested it and whether it worked would be great. List of open PRs: <https://github.com/NixOS/nixpkgs/pulls>, for more about reviewing contributions: <https://hydra.nixos.org/job/nixpkgs/trunk/manual/latest/download/1/nixpkgs/manual.html#chap-reviewing-contributions>. Reviewing isn't mandatory, but it would help out a lot and reduce the average time-to-merge for all of us. Thanks a lot if you do! -->
###### Motivation for this change

This change adds support for declarative networks with `iwd(8)`, similar
to `networking.wireless.networks` and an `interfaces` option which tells
the service which to use (by default, the interfaces will be found
automatically).

In contrast to the `wpa_supplicant` module it's still possible to
add additional networks imperatively using `iwctl(1)` since all networks
are stored as separated files in `/var/lib/iwd`.

Please note that this change doesn't add support for 802.1x networks[1]
as this would currently mean that we'd have to store sensitive secrets
for those networks world-readably in the Nix-store. It's recommended to
add those networks imperatively.

This module supports open networks and networks with a WPA PSK or a raw
passphrase. If a SSID doesn't contain alphanumeric chars and `-_` only,
the network-configuration will be stored as `=<hex-encoded-version>.<ext>`
in `/var/lib/iwd`.

[1] https://iwd.wiki.kernel.org/networkconfigurationsettings#x_wpawpa2_enterprise_settings

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

###### Notify maintainers

cc @

Ma27 committed:
nixos/iwd: add `networks` and `interfaces` option

This change adds support for declarative networks with `iwd(8)`, similar
to `networking.wireless.networks` and an `interfaces` option which tells
the service which to use (by default, the interfaces will be found
automatically).

In contrast to the `wpa_supplicant` module it's still possible to
add additional networks imperatively using `iwctl(1)` since all networks
are stored as separated files in `/var/lib/iwd`.

Please note that this change doesn't add support for 802.1x networks[1]
as this would currently mean that we'd have to store sensitive secrets
for those networks world-readably in the Nix-store. It's recommended to
add those networks imperatively.

This module supports open networks and networks with a WPA PSK or a raw
passphrase. If a SSID doesn't contain alphanumeric chars and `-_` only,
the network-configuration will be stored as `=<hex-encoded-version>.<ext>`
in `/var/lib/iwd`.

[1] https://iwd.wiki.kernel.org/networkconfigurationsettings#x_wpawpa2_enterprise_settings

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-ready-for-review-may-2019/3032/99

Ma27 committed:
iwd: add patch to avoid modification of config files managed by Nix

Ma27 commented:
As discussed with @fpletz, we should ensure that `iwd` doesn't override SSID-configs managed by Nix. I added a patch which ensures that this doesn't happen if the file to update in `/var/lib/iwd` points to a store path.

dtzWill commented on nixos/modules/services/networking/iwd.nix:
```diff
@@ -4,16 +4,141 @@ with lib;
+      description = ''
+        List of interfaces to use.
+      '';
+    };
```
Can the interfaces change be done separately?
It's not needed for the network changes at all AFAICT so it's easily separated.
The reason is iwd and interface handling, especially naming, might need some careful work.

* Interface strings: interaction and reliance on use of `usePredictableInterfaceNames`
* Implications re:iwd's `UseDefaultInterface` (and the shipped `.link` file).
* (Is there some existing interface abstraction we should be tying to?)
* The service, I think?, if used with particular interfaces maybe should: `wants`, `requires` `sys-subsystem-net-devices-${ifacename-escaped-for-systemd}` (not sure)
* iwd's help doesn't tell you, but my reading of the code in `src/manager.c` suggests the interface list is comma-separated not space-separated, FWIW

All said, what's the use for specifying interfaces explicitly anyway? Just curious :).

And the rest is exciting and I think close to merge-ready, hence the suggestion for splitting off the interface-related bits.
Thoughts?

Ma27 replied:
> All said, what's the use for specifying interfaces explicitly anyway? Just curious :).

It might be possible that I'm missing something, but iwd constantly tried to use wlan0 by default which doesn't even exist on my system.

Ma27 replied:
@dtzWill any updates here?

dtzWill commented on nixos/modules/services/networking/iwd.nix:
```diff
@@ -4,16 +4,141 @@ with lib;
+      else "string which must have at least ${toString min} and at most ${toString max} chars.";
+    inherit (types.str) merge;
+    check = x: isString x && stringLength x >= min && stringLength x <= max;
+  };
```
Wishlist: some sort of test to ensure this all works as expected--at least to catch complete breakage.  Are these details re:knownnetworks population something likely to stay the same? I suspect so, but interested if anyone has better information on this.

Ma27 replied:
Do you have a good suggestion how to test wireless stuff in a VM test? Otherwise I'm currently not sure how to implement this :/

Ma27 replied:
@dtzWill do you have an idea how to integration-test wifi connectivity? Or would it be sufficient from your PoV to check if the `known-networks` cmd detects the declaratively specified networks?

(the full integration-test may be helpful to test my additional patch btw since this testing it against new `iwd` versions always takes some time).

dtzWill commented on nixos/modules/services/networking/iwd.nix:
```diff
@@ -4,16 +4,141 @@ with lib;
+    # Cannot use the SSID here: the hex-encode is done at build-time, not eval time
+    # and several special chars in a file name would break `writeText`.
+    file = pkgs.writeText "ssid" ''
+      [Security]
```
There's some INI generation bits in `lib/generators.nix` -- would that be useful here?

dtzWill commented on nixos/modules/services/networking/iwd.nix:
```diff
@@ -22,6 +147,17 @@ in {
 
+    systemd.services.iwd.preStart = mkIf (cfg.networks != {}) (concatStrings (
+      flip mapAttrsToList predefinedNetworks (file: content: ''
+        ln -sf '${content.file}' "${encodeIfNeeded file content.extension}"
```
Maybe dumb question, but does the preStart run outside of whatever systemd bits may be limiting iwd's filesystem accesses at runtime? I'm not sure which directives are enabled anyway, but curious as offhand I don't know if that's already locked down in preStart or not.  I suppose if it works at all then that means it's good :).

Ma27 replied:
At least not yet, I'm using this change for several months and the preStart hook works fine. If the upstream systemd unit (we're currently using and modifying in the module) becomes more strict, we'd have to fix it accordingly IMHO.

dtzWill commented on pkgs/os-specific/linux/iwd/0001-agent-network_connect_psk-check-if-the-config-file-f.patch:
```diff
@@ -0,0 +1,60 @@
+ 		network->ask_passphrase ? "true" : "false");
+-
+ 	if (network->ask_passphrase) {
++		char *store_dir = getenv("NIX_STORE_DIR");
```
Perhaps this should be taken from NixOS config values instead of runtime environment?

Ma27 replied:
Is there any config option for this I'm missing?

bqv replied:
builtins.storeDir

Ma27 replied:
Fixed now.

Ma27 commented:
As I didn't find a way to mark an already reviewed PR back as draft, I added a `do no merge` note at the top. I just found a weird issue where `iwd` segfaults when I want to connect to a certain SSID (only reproducible with this SSID) and as I'm running `iwd` with these changes I'd like to make sure first that this is either unrelated or gets fixed by me :)

Ma27 commented:
Seems to be fine again with iwd 1.6 :)

bqv commented:
I've also been using this as my daily, for what it's worth. It works well and good :)

Ma27 commented:
Good to know. I'm using that as well, however I occasionally have some weird config issues I need to debug before considering this mergable.

Ma27 commented:
@bqv updated the changes to work properly with `iwd` 1.8.

I'd be glad to have a few testers before merging this (cc @dtzWill @Mic92 @bqv @teto)

Ma27 commented on pkgs/os-specific/linux/iwd/0001-agent-network_connect_psk-check-if-the-config-file-f.patch:
```diff
@@ -0,0 +1,59 @@
++{
++	return l_dbus_message_new_error(msg, IWD_SERVICE ".NotSupportNixOS",
++					"Cannot update SSID %s as it's configured in a file that points to a store path! Please update your NixOS config accordingly!", ssid);
++}
```
This should probably be added to `dbus.h` to fix `-Wimplicit-function-declaration`.

Ma27 replied:
Fixed.

bqv commented:
> @bqv updated the changes to work properly with `iwd` 1.8.
> 
> I'd be glad to have a few testers before merging this (cc @dtzWill @Mic92 @bqv @teto)

I've just updated this. No problems so far, will update if I have any later!

onny commented:
Hey, what is the best way to test this PR in a running NixOS installation on Linux?

Ma27 commented:
It's unfortunately quite hacky:

``` nix
{
  disabledModules = [ "services/networking/iwd.nix" ];
  imports = [ /path/to/git/checkout/of/this/branch/nixos/modules/services/networking/iwd.nix ];
}
```

Mic92 commented:
> Hey, what is the best way to test this PR in a running NixOS installation on Linux?

Checkout the repo. Install [hub](https://github.com/github/hub). Than do:

```
$ hub checkout https://github.com/NixOS/nixpkgs/pull/75800
```

You can also cherry pick it to master or a release branch (not sure if it applies cleanly though)

```
git checkout release-20.03
hub am https://github.com/NixOS/nixpkgs/pull/75800
```

To install it:

```
nixos-rebuild -I nixpkgs=/path/to/nixpkgs switch
```

Mic92 commented:
> It's unfortunately quite hacky:
> 
> ```nix
> {
>   disabledModules = [ "services/networking/iwd.nix" ];
>   imports = [ /path/to/git/checkout/of/this/branch/nixos/modules/services/networking/iwd.nix ];
> }
> ```

This does not apply `iwd` though.

Ma27 commented:
Didn't test it, but why actually?

onny commented:
@Mic92: Thank you this worked for me!
I was able to give this PR a try. Thanks for implementing this, it is super useful!

I tried configuring it like this:
```
wireless.iwd = {
  enable = true;
  networks = {
    "WLAN-936189".passphrase = "3603005128703121";
    "Domestic Hacking WG".passphrase = "12345678910";
  };
};
```
Which results in different files and it doesn't seem iwd is still able to autoconnect:
```
ls -la /var/lib/iwd
[...]
lrwxrwxrwx 1 root root   48 Jul 29 12:32 '=446f6d6573746963204861636b696e67205747.psk' -> /nix/store/yvwr68q7akzklsfai3qsb3x00hj5qqnn-ssid
lrwxrwxrwx 1 root root   48 Jul 29 12:36  WLAN-936189.psk -> /nix/store/kkmrqb20zaq8qw0h0mlxxy8cm9k8c2zk-ssid
```
The format inside the both profiles looks right but the essid name above looks strange. Unfortunately autoconnect does not work out of the box. I have to manually connect to the network again :(

Ma27 commented:
> Further I really would love to see 802.1x support, otherwise some networks would be missing in my configuration

As described in the PR, big :-1:. I consider 802.1x credentials to be **extremely** sensitive so they must not be in a world-readable store.

> Usually the profile would have a name like Domestic Hacking WG.psk and doesn't need the encoding

Good catch! Spaces are indeed alloewd in the filenames. Will fix this.

Ma27 commented:
@onny fixed the issue for you.

Also modified the `preStart` script to remove symlinks to `/nix/store` that aren't part of `cfg.networks` anymore.

Ma27 commented:
I'm using this for a while now (and others seem to do this as well). I'd wait a week now to check if folks experience regressions from the changes I pushed today, after that I'd consider this mergable.

cc @dtzWill @Mic92 @onny @bqv

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-already-reviewed/2617/196

Ma27 commented:
cc @Infinisil as you are fairly experienced with our module-system, would you mind taking a look at this in case you have time to? :)

Infinisil commented on nixos/modules/services/networking/iwd.nix:
```diff
@@ -4,16 +4,142 @@ with lib;
+      type = types.nullOr (types.listOf types.str);
+      example = [ "wlp2s0" ];
+      apply = x: if x == null then x else concatStringsSep " " x;
+      default = null;
```
Generally lists should be avoided as types, because they compose very badly. E.g. once you have a list of say 10 interfaces, if you want to remove one, you have to `mkForce` the 9 of them that you want which is pretty bad. Instead I suggest using `types.attrsOf types.bool`, allowing you to use `{ wlp2s0 = true; }` to enable one and `{ wlp2s0 = mkForce false; }` to disable it anywhere else again.

In addition, avoid using `apply`, since it makes reading config values for other modules tricky. E.g. in this case, to figure out whether any interface is enabled, they'd have to split `config.networking.wireless.iwd` on empty chars first.

Ma27 replied:
> Generally lists should be avoided as types, because they compose very badly. E.g. once you have a list of say 10 interfaces, if you want to remove one, you have to mkForce the 9 of them that you want, which is rather ugly

In theory I agree with your reasoning, but I don't think we should do that here: I mean, how many wireless network interfaces does an average machine have? Also, you probably have to specify those per machine(-type) either way due to the variety of different names due to udev's predictable interface-naming feature (which we have enabled by default).

I think that for such a small value, the (IMHO more intuitive) `listOf`type is preferrable.

bqv replied:
As this isn't something I forsee anyone overriding anywhere but the same place it's set, I think it's perfectly reasonable to leave it as a list, though I also agree generally with the idea

Infinisil replied:
Imo making this an `attrsOf bool` instead isn't a big cost, and it would set a good sample for other modules.

But also, a problem I haven't mentioned yet: `listOf` arguably breaks determinism of the module system, because it makes the order of elements depend on the order of module imports: Changing `imports = [ ./foo.nix ./bar.nix ];` doesn't have to evaluate to the same thing as `imports = [ ./bar.nix ./foo.nix ];` for `listOf` options. I previously had to fix a bug due to this with https://github.com/NixOS/nixpkgs/pull/77088.

I think `listOf` should really only be used when it's a single list value that's assigned, e.g. `services.xserver.synaptics.buttonsMap`. And I actually plan to make an RFC to slowly deprecate `listOf` because of the problems I mentioned above.

Ma27 replied:
> Imo making this an attrsOf bool instead isn't a big cost, and it would set a good sample for other modules.

I don't know tbh: I personally think that this interface is a little bit too clumsy.

> But also, a problem I haven't mentioned yet: listOf arguably breaks determinism of the module system

OTOH you sometimes have to depend on the order of values in a list which doesn't work anymore with attrs as those are always ordered alphabetically, right? Anyway, I don't see why this issue applies here.

Infinisil replied:
Usually the order doesn't matter yeah. But it still messes with determinism a bit, since flipping imports can lead to different final system derivations.

Anyways, I have a plan for fixing these problems, let's leave this at `listOf str` for now then :)

Infinisil commented on nixos/modules/services/networking/iwd.nix:
```diff
@@ -4,16 +4,142 @@ with lib;
+        "karlsruhe.freifunk.net" = {};
+        "secured".passphrase = "12345678";
+        "network-with-psk".psk = "bf4c086e055e492373a0f59ff614f61fc2d597a56db94e17a165f9ef1d42c066";
+      };
```
For examples to render on multiple lines you need to wrap it in a string and use `lib.literalExample`:
```nix
{
  example = lib.literalExample ''
    {
      "karlsruhe.freifunk.net" = {};
      # ...
    }
  '';
}
```

Ma27 replied:
Good point! :)

Infinisil commented on nixos/modules/services/networking/iwd.nix:
```diff
@@ -4,16 +4,142 @@ with lib;
+      then "string with exactly ${toString min} chars."
+      else "string which must have at least ${toString min} and at most ${toString max} chars.";
+    inherit (types.str) merge;
+    check = x: isString x && stringLength x >= min && stringLength x <= max;
```
Since this type would be useful for many other modules, I suggest adding it to `lib/types.nix`. I'd also name it `lengthCheckedString` instead. Note that `name` should also be set to what the type variable is named when doing this.

Infinisil commented on nixos/modules/services/networking/iwd.nix:
```diff
@@ -4,16 +4,142 @@ with lib;
+    allowedChars = stringToCharacters "01234556789abcdefghijklmnopqrstuvwxyz-_ ";
+    skipEncoding = all (char: elem (toLower char) allowedChars) (stringToCharacters file);
+  in
+    "/var/lib/iwd/${if skipEncoding then file else "$(${encodeFileName} \"${file}\")"}.${ext}";
```
Since nothing actually needs these values at eval time, you can instead do this check during runtime in `preStart` directly.

Infinisil commented on nixos/modules/services/networking/iwd.nix:
```diff
@@ -4,16 +4,142 @@ with lib;
+    # Cannot use the SSID here: the hex-encode is done at build-time, not eval time
+    # and several special chars in a file name would break `writeText`.
+    file = pkgs.writeText "ssid" (generators.toINI {} ({
+      Settings.AutoConnect = cfg.autoConnect;
```
You can also do this at runtime by dropping the use of `generators.toINI` and manually constructing the INI file

Ma27 replied:
Well, I'm not sure whether we should do this: I check if a file in `/var/lib/iwd` is a store-path and if that's the case, IWD now refuses to modify/remove/whatever this network configuration. This wouldn't be possible anymore with runtime-construction and thus leading to eventual consistency as `iwd` can simply modify the network's configuration and therefore breaking the immutability-goal of NixOS.

Infinisil replied:
It's still possible to do this, which imo is even nicer since it doesn't rely on checking for Nix store symlinks: Write the list of files that were deployed declaratively to `/var/lib/iwd/.declarative-nixos-networks` after linking them. Then in the next `preStart`, remove all the files given in that list first.

Ma27 replied:
I like the idea and will give it a try soon :)

Infinisil commented on nixos/modules/services/networking/iwd.nix:
```diff
@@ -4,16 +4,142 @@ with lib;
+            '';
+          };
+
+          passphrase = mkOption {
```
And manually constructing the INI file during runtime has the benefit that you can convert the `psk` and `passphrase` options to `pskFile` and `passphraseFile` respectively, which then won't force the user into putting their password in the Nix store. When constructing the INI file in `preStart` you can read these file paths to get the secret values.

Note that if people don't care about this, they can still do `psk = pkgs.writeFile "foo-psk" "the-psk"`, which still puts the secret in the store.

Furthermore, 802.1x support can be added similarly, as requested by @onny.

Infinisil replied:
Relevant issue: https://github.com/NixOS/nixpkgs/issues/24288

Infinisil commented on nixos/modules/services/networking/iwd.nix:
```diff
@@ -4,16 +4,142 @@ with lib;
+          autoConnect = mkOption {
+            default = true;
+            type = types.bool;
+            apply = boolToString;
```
`apply` should also be avoided here for the same reason (and similarly for the `hidden` option)

Infinisil commented on nixos/modules/services/networking/iwd.nix:
```diff
@@ -22,8 +148,29 @@ in {
+      find /var/lib/iwd -type l -lname '${builtins.storeDir}/*' ${optionalString (ssids != {}) ''
+        -not \( ${concatMapStringsSep " -o " ({ local, ... }: "-name '${baseNameOf local}*'") ssids} \) \
+      ''} \
+        | xargs -I {} rm -f {}
```
This seems really complicated and I'm having a hard time grasping what it actually does. I think some comments would help a lot with that. You could also use `find`'s `-exec` option to simplify this a bit.

Mic92 replied:
If you just want to delete stuff you can also use find's builtin `-delete`.

Ma27 replied:
To summarize, this removes all files in `/var/lib/iwd` that are a symlink to a store-path, but **aren't** defined in `cfg.networks`.

Ma27 committed:
nixos/iwd: refactor options and cleanup script

Ma27 committed:
types/lengthCheckedString: init

New type which can be used to define an option that expects a
`str` with a declarable length. This originates from the `iwd`-module
where this was used to validate the length of a wireless network's PSK.

It can be used like this:

   # Option expecting a string that must have at least 5 chars and
   # at most 10 chars.
   mkOption { /* ... */
     type = types.lengthCheckedString 5 10;
   }

If a string can only have a single specified length, it can be declared
like this:

   # Option expecting a string that must have exactly 23 chars.
   mkOption { /* ... */
     type = types.fixedLengthString 23;
   }

Ma27 commented:
@Infinisil thanks a lot for the detailedreview. I fixed most of your notes and left a comment re `cfg.interfaces` and runtime construction for network configs as I'm not sure whether this is a good idea here (but I'm happy to discuss though).

Ma27 commented:
After giving this a lot of thought, I decided to close this now. In fact, I started an attempt to implement the new suggestions this week and spent a lot of time & energy with my colleague @lheckemann last night to patch the C code of `iwd` for this PR.

First of all, let me explain why I'm stopping this effort now: unfortunately `iwd` assumes that it's the only process which is doing *anything* in `/var/lib/iwd` and patching the code to make sure that `iwd` doesn't touch networks defined by NixOS is fairly hard (e.g. removing known networks isn't supported here atm). Also, `iwd` has some fairly weird behavior when establishing a connection (it wasn't reproducible when our code-path was executed and when it wasn't - we spent pretty much time in `gdb` on that).

Apart from that I'm losing the interest in this change as I'm currently re-evaluating `wpa_supplicant` for my personal setup as `iwd` repeatedly has trouble to log into wpa2-enterprise networks (such as the wifi in my office).

However, if someone wants to continue working on this, that's perfectly fine. The current diff on my `nixpkgs` checkout can be found in https://gist.github.com/Ma27/fc3e2bfcdd0aba7e65552a97189d7715 and Linus added some more improvements today in https://git.sr.ht/~linuxhackerman/iwd/log.

Last but not least, @Infinisil & @bqv thanks for the review. I'll check if your ideas are somehow applicable for `networking.wireless.networks` and would file a new PR for that if I get a result that can be included into upstream nixpkgs :)

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/iwd-declarative-wifi-configs-networks-credentials/11267/2
